heat/core/tests/test_arithmetics.py

```diff
@@ -98,44 +108,154 @@ def test_add(self):
+        self.assertTrue(torch.equal(self.a_tensor.larray, underlying_torch_tensor))
+        self.assertTrue(ht.equal(self.another_tensor, ht.array([[2.0, 2.0], [2.0, 2.0]])))
+        self.a_tensor.larray = ht.array([[1.0, 2.0], [3.0, 4.0]]).larray
+        underlying_torch_tensor = self.a_tensor.larray
+
+        self.assertTrue(ht.equal(ht.add_(self.a_tensor, self.a_vector), result))
+        self.assertTrue(ht.equal(self.a_tensor, result))
+        self.assertTrue(torch.equal(self.a_tensor.larray, underlying_torch_tensor))
+        self.assertTrue(ht.equal(self.a_vector, ht.float32([2, 2])))
+        self.a_tensor.larray = ht.array([[1.0, 2.0], [3.0, 4.0]]).larray
+        underlying_torch_tensor = self.a_tensor.larray
+
+        self.assertTrue(ht.equal(ht.add_(self.a_tensor, self.an_int_scalar), result))
+        self.assertTrue(ht.equal(self.a_tensor, result))
+        self.assertTrue(torch.equal(self.a_tensor.larray, underlying_torch_tensor))
+        self.assertTrue(ht.equal(self.an_int_scalar, 2))
+        self.a_tensor.larray = ht.array([[1.0, 2.0], [3.0, 4.0]]).larray
+        underlying_torch_tensor = self.a_tensor.larray
+
+        self.assertTrue(ht.equal(ht.add_(self.a_split_tensor, self.a_tensor), result))
+        self.assertTrue(ht.equal(self.a_split_tensor, result))
+        self.assertTrue(torch.equal(self.a_split_tensor.larray, underlying_split_torch_tensor))
+        self.assertTrue(ht.equal(self.a_tensor, ht.array([[1.0, 2.0], [3.0, 4.0]])))
+        self.a_split_tensor.larray = ht.array([[2.0, 2.0], [2.0, 2.0]]).copy().resplit_(0).larray
+        underlying_split_torch_tensor = self.a_split_tensor.larray
+
+        # Single element split
+        a = ht.array([1, 2], split=0)
+        b = ht.array([1], split=0)
+        a += b
+        self.assertTrue(ht.equal(a, ht.array([2, 3])))
+        if a.comm.size > 1:
+            if a.comm.rank < 2:
+                self.assertEqual(a.larray.size()[0], 1)
+            else:
+                self.assertEqual(a.larray.size()[0], 0)
+
+        # test with differently distributed DNDarrays
+        a = ht.ones(10, split=0)
+        b = ht.zeros(10, split=0)
+        a = a[:-1]
+        a_lshape = a.lshape
+        a += b[1:]
+        self.assertTrue((a == 1).all())
+        self.assertTrue(a.lshape == a_lshape)
+
+        # test unbalanced
+        a = ht.ones(10, split=0)  # reset
+        a = a[1:-1]
+        a_lshape = a.lshape
+        a += b[1:-1]
+        self.assertTrue((a == 1).all())
+        self.assertTrue(a.lshape == a_lshape)
+
+        # test one unsplit
+        a = ht.ones(10, split=None)
+        b = ht.zeros(10, split=0)
+        a = a[:-1]
+        a += b[1:]
+        self.assertTrue((a == 1).all())
+        self.assertEqual(a.lshape, b[1:].lshape)
+
+        a = ht.ones(10, split=None)  # reset
+        a = a[1:]
+        a += b[:-1]
+        self.assertTrue((a == 1).all())
+        self.assertEqual(a.lshape, b[:-1].lshape)
```

Comment:
`add_` has the same property as `add` here: When adding two DNDarrays of which one has split axis `None` and the other has split axis `0`, the result has split axis `0`.

Reply:
The result, however, has new tensors internally.

```python
import heat as ht

a = ht.array([1,2,3,4], split=None)
b = ht.array([1,2,3,4], split=0)
a_ten = a.larray
a += b
print(a is a_ten)
```

```console
[1,0]<stdout>:False
[1,3]<stdout>:False
[1,1]<stdout>:False
[1,2]<stdout>:False
```

Reply:
Okay, got it. Thank you. Please confirm whether my following thought is correct: 
Whenever I resplit a DNDarray, I initialize new underlying tensors.
```
def test_split(self):
    a_tensor = ht.array([[1.0, 2.0], [3.0, 4.0]], split=None)
    underlying_torch_tensor = a_tensor.larray
    a_tensor.resplit_(0)
    self.assertIs(a_tensor.larray, underlying_torch_tensor)  # test in-place
```
```
======================================================================
FAIL: test_split (heat.core.tests.test_arithmetics.TestArithmetics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sche_lk/heat/heat/core/tests/test_arithmetics.py", line 115, in test_split
    self.assertIs(a_tensor.larray, underlying_torch_tensor)  # test in-place
AssertionError: tensor([[1., 2.]]) is not tensor([[1., 2.],
        [3., 4.]])

----------------------------------------------------------------------
Ran 55 tests in 0.558s

FAILED (failures=1)
ok

======================================================================
FAIL: test_split (heat.core.tests.test_arithmetics.TestArithmetics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sche_lk/heat/heat/core/tests/test_arithmetics.py", line 115, in test_split
    self.assertIs(a_tensor.larray, underlying_torch_tensor)  # test in-place
AssertionError: tensor([[3., 4.]]) is not tensor([[1., 2.],
        [3., 4.]])
```
Therefore, in the tested case above the addition with `add_` is not actually in-place. To make it in-place the behaviour of `add_` has to differ from `add` here so that the split axis of _a_ doesn't change as you mentioned intially.

Reply:
Many functions that change the distribution create new tensors to fit the new local shapes.

We can discuss in a larger group next time what we want to do when the axes differ. Personally, I would prefer to be stricter on the arithmetic in-place operations.